thundra/thundra_agent.py

```diff
@@ -142,70 +146,66 @@ def wrapper(event, context):
 
 
-    def wait_for_debugger(self):
-        start = time.time()
-        prev_rchar = 0
-        prev_wchar = 0
-        while time.time() - start < config.debugger_max_wait_time()/1000:
-            try:
-                io_metrics = self.get_debugger_proxy_io_metrics()
-                if not io_metrics:
-                    time.sleep(1)
-                    break
-                if prev_rchar != 0 and prev_wchar != 0 and ( io_metrics[0] == prev_rchar and io_metrics[1] == prev_wchar):
+    def start_debugger_tracing(self):
+        try:
+            import ptvsd
+            ptvsd.tracing(True)
+
+            ptvsd.enable_attach(address=("localhost", config.debugger_port()))
+            if not self.debugger_process:
+                env = os.environ.copy()
+                env['BROKER_HOST'] = str(config.debugger_broker_host())
+                env['BROKER_PORT'] = str(config.debugger_broker_port())
+                env['DEBUGGER_PORT'] = str(config.debugger_port())
+                env['AUTH_TOKEN'] = str(config.debugger_auth_token())
+                env['SESSION_NAME'] = str(config.debugger_session_name())
+                context = self.plugin_context["context"]
+                if hasattr(context, 'get_remaining_time_in_millis'):
+                    env['SESSION_TIMEOUT'] = str(context.get_remaining_time_in_millis() + int(time.time()*1000.0))
+                self.debugger_process = subprocess.Popen(["python", "thundra/debug/bridge.py"], stdout=subprocess.PIPE, stdin=subprocess.PIPE, env=env)
+
+            start_time = time.time()
+
+            while time.time() < (start_time + config.debugger_max_wait_time()/1000) and not ptvsd.is_attached():
+                if self.debugger_process.poll() is None:
+                    ptvsd.wait_for_attach(0.01)
+                else:
                     break
-                prev_rchar = io_metrics[0]
-                prev_wchar = io_metrics[1]
-                time.sleep(1)
 
-            except Exception as e:
-                logger.error("error while waiting for proxy process: {}".format(e))
+            if not ptvsd.is_attached():
+                ptvsd.tracing(False)
+            else:
+                ptvsd.tracing(True)
+
+        except Exception as e:
+            logger.error("error while setting tracing true to debugger using ptvsd: {}".format(e))
 
     def stop_debugger_tracing(self):
         try:
             import ptvsd
             ptvsd.tracing(False)
-            self.wait_for_debugger()
+            from ptvsd.attach_server import debugger_attached
+            debugger_attached.clear()
         except Exception as e:
             logger.error("error while setting tracing false to debugger using ptvsd: {}".format(e))
 
         try:
             if self.debugger_process:
-                self.debugger_process.kill()
-                self.debugger_process.wait()
+                o, _ = self.debugger_process.communicate(b"fin\n")
```

Comment:
should we also kill process? where we are killing `debugger_process`?

Reply:
ok, got it

Reply:
I think, after sending `fin` message, it might be better to wait `debugger_process` to terminate by `debugger_process.wait()`.